janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/MultiPageEntryBuffer.java

```diff
@@ -0,0 +1,406 @@
+    private static int INITIAL_CAPACITY = 2;
+
+    private static final Logger log = LoggerFactory.getLogger(BufferPageUtils.class);
+
+    //NOTE: tracking pages in a list, which internally maintains an array, creates additional overhead;
+    // if we maintain our page list in an array directly, this will save this overhead at the cost of some more boilerplate code here
+    // However in practice the number of multi-page buffers is usually relatively small, so at the moment it seems to be
+    // a reasonable price for having less code
+    private ArrayList<BufferPage> pages;
+
+    public MultiPageEntryBuffer(BufferPage initialPage) {
+        this.pages = new ArrayList<>(INITIAL_CAPACITY);
+        if (initialPage instanceof SinglePageEntryBuffer) {
+            pages.add(new BufferPage(initialPage.getOffsetIndex(), initialPage.getRawData()));
+        } else
+            pages.add(initialPage);
+    }
+
+    MultiPageEntryBuffer(List<BufferPage> allPages) {
+        this.pages = new ArrayList<>(allPages);
+    }
+
+    @Override
+    public int numPages() {
+        return pages.size();
+    }
+
+    @Override
+    public int numEntries() {
+        if (pages == null) {
+            return 0;
+        }
+
+        int numEntries = 0;
+
+        for (BufferPage p : pages) {
+            numEntries += p.numEntries();
+        }
+        return numEntries;
+    }
+
+    @Override
+    public int byteSize() {
+        if (pages == null) {
+            return 0;
+        }
+
+        int byteSize = 0;
+        for (BufferPage p : pages) {
+            byteSize += p.byteSize() + 16;
+        }
+        return byteSize;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return pages.isEmpty();
+    }
+
+    private int getPageIndex(StaticBuffer column) {
+        //binary search for correct page on column name
+        int low = 0;
+        int high = pages.size() - 1;
+
+        while (low <= high) {
+            int mid = (low + high) >>> 1;
+
+            BufferPage midPage = pages.get(mid);
+
+            int idx = midPage.getIndex(column);
+
+            if (idx >= 0) {
+                return mid; //key found in this page
+            }
+
+            //key not found, idx contains nearest index, negated
+            idx = (-idx - 1);
+
+            if (idx <= 0) //key is smaller than the first key in this page - search previous pages
+            {
+                high = mid - 1;
+            } else if (idx >= midPage.numEntries()) //key is bigger than the last in this page - search following pages
+            {
+                low = mid + 1;
+            } else //key was not found in this page, but the nearest key found is in this page
+            {
+                return mid;
+            }
+        }
+        return -(low + 1);  //key not found in any of the pages, and is outside the range of current pages - return the nearest page
+    }
+
+    @Override
+    public EntryList getSlice(KeySliceQuery query) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'getSlice(KeySliceQuery)' has an NPath complexity of 2364](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=3231392)